### PR TITLE
fix: properly ignore source-code field in metadata linter

### LINKED ISSUE #5794: Can't ignore source-code metatada key in linter configuration
### Bug Description

With the new metadata linter that was added in #5610 and deployed with the 8.12.0 release, we now get warnings on missing metadatata keys in our `snapcraft.yaml`. As part of this feature, you can specify ignores for the linter. As my company is using snaps for in house deployment on our IoT fleet, most keys do not make sense for us so I added them to the ignore configuration. This works for all keys but the `source-code` key!

### To Reproduce

If I specify this in `snapcraft.yaml`:

```yaml
lint:
  ignore:
    - metadata:
      - source-code
```
I would expect it to suppress warnings on missing the `source-code` key. But this does not work

**I found it _does_ work to specify the key as `source_code`, with underscore instead of hyphen**, which is unexpected.

### Environment

Ubuntu 24.04 LTS, snapcraft 8.12.0

### snapcraft.yaml

```yaml
name: snapps
base: core24
version: '0.1'
summary: Single-line elevator pitch for your amazing snap # 79 char long summary
description: your description wants to look good in the snap store.

grade: stable
confinement: devmode

parts:
  my-part:
    plugin: nil

lint:
  ignore:
    - metadata:
      - contact
      - donation
      - title
      - issues
      - source-code
      - website
      - license
```

### Relevant log output

```shell
$ snapcraft pack
Lint information:
- metadata: Metadata field 'source-code' is missing or empty. (https://documentation.ubuntu.com/snapcraft/stable/reference/project-file/snapcraft-yaml/#source-code)
Packed snapps_0.1_amd64.snap
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by bepri**
- [x] Have you followed the [guidelines for contributing](https://github.com/canonical/snapcraft/blob/main/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] Have you successfully run `make lint`?
- [x] Have you successfully run `make test`?

---

Fixes #5794. Does a little fly-by typing work while I'm in this file, as well.

### CODE REVIEW

**Comment by mbeijen** [comment]
thanks, I built snapcraft from this branch and can confirm it fixes my issue!

File: tests/unit/linters/test_metadata_linter.py [hunk 1]
```diff
@@ -142,6 +144,75 @@ def test_metadata_linter_ignore_some_issues(new_dir, snap_yaml_data):
...
+def test_metadata_linter_source_code(
+    new_dir: Path,
+    snap_yaml_data: Callable[..., dict[str, Any]],
+    lint_source: bool,
+    specify_source: bool,
+    linter_issue: LinterIssue | None,
+) -> None:
+    """Make sure that the source-code field is handled properly.
+
+    https://github.com/canonical/snapcraft/issues/5794"""
```
Comment (upils) [inline tests/unit/linters/test_metadata_linter.py:198]: suggestion: I understand the root cause of this issue was a typo and not a subtle bug requiring a subtle fix, so I am not sure referencing the issue is really necessary.

File: tests/unit/linters/test_metadata_linter.py [hunk 2]
```diff
@@ -142,6 +144,75 @@ def test_metadata_linter_ignore_some_issues(new_dir, snap_yaml_data):
...
+            LinterIssue(
+                name="metadata",
+                result=LinterResult.IGNORED,
+                text="Metadata field 'source-code' is ignored.",
+            ),
+            id="skip-ignored",
+        ),
+    ],
+)
+def test_metadata_linter_source_code(
```
Comment (upils) [inline tests/unit/linters/test_metadata_linter.py:189]: Unless I am missing something it looks like this test is testing some code paths already tested in `test_metadata_linter_missing_metadata` and `test_metadata_linter_ignore_all_issues`. Are these edge cases linked to the `source-code` field?

File: tests/unit/linters/test_metadata_linter.py (hunk 1 above)
Comment (bepri) [reply tests/unit/linters/test_metadata_linter.py:198]: It was a bit more subtle - the attribute needed to be referenced as `source-code` for ignoring and friendly printing purposes, but internally it gets normalized to `source_code`. That's why I wanted to put so much into the test here

File: tests/unit/linters/test_metadata_linter.py (hunk 2 above)
Comment (bepri) [reply tests/unit/linters/test_metadata_linter.py:189]: Yes - initially I almost included it in a previous test, but the parametrization made the test overly complex. I liked this "contained" behavior more, especially since this is a regression test. All of these steps are necessary to test the fix, as I need to ensure that it both understands the `source-code` lint, but understands to check the `source_code` (transformed by Pydantic) field for the lint.

File: tests/unit/linters/test_metadata_linter.py (hunk 1 above)
Comment (upils) [reply tests/unit/linters/test_metadata_linter.py:198]: Ohh... I missed that.
